src/core.jl

```diff
@@ -19,3 +19,157 @@ immutable ImageEdge
     index2::Int
     weight::Float64
 end
+
+"""
+    G, vert_map = region_adjacency_graph(seg, [weight_fn])
+
+Constructs a region adjacency graph from the `SegmentedImage`. It returns the RAG
+along with a label->vertex map. Optionally, a weight function `weight_fn` might be
+provided to set the edge weights. `weight_fn` takes two segment means and returns
+the weight of the connecting edge.
+
+"""
+function region_adjacency_graph{T<:SegmentedImage}(s::T, weight_fn::Function = default_diff_fn)
+
+    function neighbor_regions!{T<:SegmentedImage}(s::T, I::CartesianIndex, visited::AbstractArray, n::Set{Int})
+        R = CartesianRange(indices(s.image_indexmap))
+        I1, Iend = first(R), last(R)
+        t = Vector{CartesianIndex{ndims(visited)}}()
+        push!(t, I)
+
+        while !isempty(t)
+            temp = pop!(t)
+            visited[temp] = true
+            for J in CartesianRange(max(I1, temp-I1), min(Iend, temp+I1))
+                if s.image_indexmap[temp] != s.image_indexmap[J]
+                    push!(n,s.image_indexmap[J])
+                elseif !visited[J]
+                    push!(t,J)
+                end
+            end
+        end
+        n
+    end
+
+    local visited   = similar(dims->fill(false,dims), indices(s.image_indexmap))    # Array to mark the pixels that are already visited
+    G               = SimpleWeightedGraph()                                         # The region_adjacency_graph
+    vert_map        = Dict{Int,Int}()                                               # Map that stores (label, vertex) pairs
+
+    # add vertices to graph
+    add_vertices!(G,length(s.segment_labels))
+
+    # setup `vert_map`
+    for (i,l) in enumerate(s.segment_labels)
+        vert_map[l] = i
+    end
+
+    # add edges to graph
+    for p in CartesianRange(indices(s.image_indexmap))
+        if !visited[p]
+            local n = Set{Int}()
+            neighbor_regions!(s, p, visited, n)
+            for i in n
+                add_edge!(G, vert_map[s.image_indexmap[p]], vert_map[i], weight_fn(s.segment_means[s.image_indexmap[p]], s.segment_means[i]))
+            end
+        end
+    end
+    G, vert_map
+end
+
+
+"""
+    new_seg = remove_segment(seg, label, [weight_fn])
+
+Removes the segment having label `label` and returns the new `SegmentedImage`.
+For more info, see [`remove_segment!`](@ref)
+
+"""
+rem_segment{T<:SegmentedImage}(s::T, args...) = rem_segment!(deepcopy(s), args...)
+
+"""
+    remove_segment!(seg, label, [weight_fn])
+
+Removes the segment having label `label` in place, replacing it with the neighboring
```

Comment:
Separation between "removes" and "in place" is a bit confusing, Maybe "In-place removal of the ..."?